src/lib/onesource.cpp

```diff
@@ -879,6 +879,21 @@ void onesource::generatePDF(vector<SED *> &fulllib, const vector<size_t> &va,
   colAnalysis = ((fltColRF[0] >= 0) && (fltColRF[1] >= 0) &&
                  (fltColRF[2] >= 0) && (fltColRF[3] >= 0) && (fltREF >= 0));
 
+  // Prepare direct references for all variables stored inside slow containers
```

Comment:
I don't really know what this line means in terms of the modern C++ language semantics. The way I read this is that a stack-aallocated dynamic array is allocated (with `alloca()`) and then an initializer-list constructor is called with an incomplete (smaller than the full size) initializer-list, so for me this means different results (only the first element is zeroed). But I'd need to read the C++ manual to find out what this really does and if the incomplete initializer-list would really be expanded to the dynamic array size every time the method is called.

My suggestion for these variables is actually quite different, which would be to use `std::vector`s instead, benefiting from the constructor that fills the initial values with zeros and, if necessary, return the raw pointer with `std::vector::data` (which might not even be necessary).

Reply:
Humm you lost me here.... The way I read it is that `double PDFzloc[pdfmap[11].size()]` declares a C-style array PDFzloc of size pdfmap[11].size(), and the subsequent loop fills all the values with 0. A modern C++ version is std::array, that for some reason we did not use : https://stackoverflow.com/questions/30263303/what-are-the-advantages-of-using-stdarray-over-c-style-arrays

Reply:
Not exactly, that line in C would allocate a VLA (variable length array) and can't be initialized, except with an empty initializer:

```
abcd.c:9:23: error: variable-sized object may not be initialized except with an empty initializer
    9 |         double V[s] = { 0. };

```

However, in C++ that line compiles and does the expected thing: I've just looked at the binary executable and a `memset()` is inserted to fill the whole vector, even when `s` is not constant. So the original idea works:

```
double PDFzloc[pdfmap[11].size()] = {0.};
```

`std::array` can't be used because the array size is must be a compile-time constant expression, but in this case it's a variable.

https://stackoverflow.com/questions/17332360/initializing-variable-length-array

Reply:
Bad news, VLA initialization in C++ is also forbidden. The support in g++ is non-standard. To see the error, compile with `-pedantic`:

```
[hdante@host1 tmp]$ c++ -o abcd abcd.cc -O3 -g -pedantic -Werror
abcd.cc: In function ‘int main()’:
abcd.cc:11:16: error: ISO C++ forbids variable length array ‘V’ [-Werror=vla]
   11 |         double V[s] = { 0. };
      |                ^
cc1plus: all warnings being treated as errors
```

https://stackoverflow.com/a/15013295/1797000

Reply:
Since the error says that VLA themselves are forbidden in C++, now I recommend changing the VLAs to `std::vector`s.

Reply:
agreed on this conclusion then. And actually we can do `vector<double> PDFzloc(pdfmap[11].size(), 0.);` if I am not mistaken

Reply:
Hmm, actually I ran into a problem : if I use a vector I run into a problem with the reduction : 

> vector<double> PDFzloc(pdfmap[11].size(), 0.);

yields the following compilation error : 

>      /home/cohen/lsst/codes/pipelines/photoz/lephare_hdante/src/lib/onesource.cpp:906:13: error: user defined reduction not found for ‘PDFzloc’
        906 |         + : PDFzloc, PDFzqloc, PDFmassloc, PDFSFRloc, PDFsSFRloc, PDFAgeloc) \
            |             ^~~~~~~

Reply:
ok so the reduction directive needs to be defined for vectors.  In the end I get the code committed in https://github.com/hdante/lephare/commit/a236c1744174154af13622aea9740a26b69da5cd

Reply:
damn it segfault in CI and not on my system.....

Reply:
I will revert and merge the branch with your modifs, then I will open a new issue to keep track of the VLA issue

Reply:
Ok, I saw the patch was merged, thanks.